renku/core/session/notebook_service.py

```diff
@@ -0,0 +1,386 @@
+from pathlib import Path
+from time import sleep
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+from yaspin import yaspin
+
+from renku.command.command_builder import inject
+from renku.core import errors
+from renku.core.interface.client_dispatcher import IClientDispatcher
+from renku.core.management.client import LocalClient
+from renku.core.plugin import hookimpl
+from renku.core.session.utils import get_renku_project_name
+from renku.core.util import communication, requests
+from renku.core.util.git import get_remote
+from renku.core.util.urls import parse_authentication_endpoint
+from renku.domain_model.session import ISessionProvider, Session
+
+
+@inject.autoparams("client_dispatcher")
+def _get_token(client_dispatcher: IClientDispatcher, renku_url: str) -> Tuple[str, bool]:
+    """Get a token for authenticating with renku.
+
+    If the user is logged in then the JWT token from renku login will be used.
+    Otherwise the anonymous user token will be used. Returns the token and a flag to
+    indicate if the user is registered (true) or anonymous(false).
+    """
+    client = client_dispatcher.current_client
+    registered_token = client.get_value(section="http", key=urllib.parse.urlparse(renku_url).netloc)
+    if not registered_token:
+        return _get_anonymous_credentials(renku_url=renku_url), False
+    return registered_token, True
+
+
+@inject.autoparams("client_dispatcher")
+def _get_anonymous_credentials(client_dispatcher: IClientDispatcher, renku_url: str) -> str:
+    def _get_anonymous_token() -> Optional[str]:
+        import requests
+
+        with requests.Session() as session:
+            url = urllib.parse.urljoin(renku_url, "api/user")
+            try:
+                session.get(
+                    url,
+                    headers={
+                        "X-Requested-With": "XMLHttpRequest",
+                        "X-Forwarded-Uri": "/api/user",
+                    },
+                )
+            except (requests.exceptions.RequestException, requests.exceptions.ConnectionError):
+                pass
+            return session.cookies.get("anon-id")
+
+    client = client_dispatcher.current_client
+    renku_host = urllib.parse.urlparse(renku_url).netloc
+    anon_token = client.get_value(section="anonymous_tokens", key=renku_host)
+    if not anon_token:
+        anon_token = _get_anonymous_token()
+        if not anon_token:
+            raise errors.AuthenticationError(
+                "Could not get anonymous user token from Renku. "
+                f"Ensure the Renku deployment at {renku_url} supports anonymous sessions."
+            )
+        client.set_value(section="anonymous_tokens", key=renku_host, value=anon_token, global_only=True)
+    return anon_token
+
+
+@inject.autoparams()
+def _get_client(client_dispatcher: IClientDispatcher) -> LocalClient:
+    """Get the local client."""
+    client = client_dispatcher.current_client
+    return client
```

Comment:
I had some issues with this but I managed to get by it. Hopefully the implementation is good enough. With or without the Proxy trying to inject in a class variable did not work. This is because the NotebookServiceSessionProvider is initialized when registered as a plugin and this causes the injection to occur when it is too early and fail.

Using a class method bypasses this.